jwks.go

```diff
@@ -154,8 +167,10 @@ func (j *JWKS) ReadOnlyKeys() map[string]interface{} {
 
 // getKey gets the jsonWebKey from the given KID from the JWKS. It may refresh the JWKS if configured to.
 func (j *JWKS) getKey(kid string) (jsonKey interface{}, err error) {
+	const useEncryption = "enc"
```

Comment:
I think I'd like to see this string as an exported constant of a new type `JWKUse`. We could also add another constant `UseSignature`.

I think it would be a good idea to make a new field on the `keyfunc.Option` data structure for `[]JWKUse`. When the slice has a length of `1` or more, only those values are accepted and the rest are ignored. When that slice has a length of `0`, it would use the default behavior. This would be important so that users of this project could configure behavior for their use case without modifying any project code. 

I'm not sure about the default behavior currently. I think there are three reasonable mutually-exclusive options:
1. Accept any `use` value.
2. Only accept `use` values of `sig`.
3. Reject all `use` values of `enc`.

I'm leaning towards option `1` because there is no change of introducing backwards incompatible behavior and it would also be forwards compatible if the upstream project `github.com/golang-jwt/jwt/v4` decided to include support for JWT payload encryption/decryption (and somehow a `jwt.Keyfunc` was responsible for that).

Reply:
> if the upstream project github.com/golang-jwt/jwt/v4 decided to include support for JWT payload encryption/decryption (and somehow a jwt.Keyfunc was responsible for that)

I find this very unlikely, perhaps the best option is `3`, which is the current behavior of this PR.

Reply:
It depends on if this project will have applications beyond plugging into go-jwt. 

If the goal was to be a general purpose JWKS package that supported any type of key usage then the consumer of the package would probably want the formatted public key (underlying go crypto struct of particular key type) in addition to the metadata associated with the key (`use`, `alg`, `x5c`, ...). This would allow more complex logic such as pairing up a key to JWT/JWE with their allowed usage, algorithm, or maybe verifying the `x5c` certificate chain. Right now the only thing consumers get out of the package is the formatted public key. I personally don't mind that this project is limited in scope.

For keeping this package's compatibility we could go with your suggestion of keeping a set of allowed key usages. Empty for allow all types. Defaulting to allow all would keep the same functionality. Personally using this package I would configure my allowed usage as `allowedKeyUsage := []string{"", "sig"}`. We could add a note to the project readme about choosing this optional value.

Reply:
> If the goal was to be a general purpose JWKS package

I'm currently creating such a package in a separate project: https://github.com/MicahParks/jwkset. The current purpose of the linked project is to be a server-side implementation for a JWK Set, but I have long term plans to add a generalized JWK Set client to that repository. My current preference is to keep this project, `keyfunc`, as a limited scope project tied to the upstream JWT package `github.com/golang-jwt/jwt/v4`.

> maybe verifying the `x5c` certificate chain

I haven't considered this until now, I think this has potential to be a good new GitHub issue.

On the question of default behavior, what would you say is your preference, @trevorlyman? I see how you've mentioned the 100% backward compatible way of accepting any `use` value. But I also see that the first commits in this PR accepted everything except a `use` value of `enc`. I can see the argument for the exclusion and it might not be considered a backwards-incompatible change, since [an RFC dictates](https://www.rfc-editor.org/rfc/rfc7517.html#section-4.2) it should be used for encryption (implying it shouldn't be used for signatures).

Reply:
If this package is committed to supporting go-jwt signatures as it's primary function, then to match the RFC I believe we should **not** be allowing `"use": "enc"` by default. It's a somewhat obscure security risk, but a key should only be used for it's authorized purpose.